### DISCUSSION TITLE
How to use invoke to perform synchronous logic in an actor

### DISCUSSION THREAD

**Discussion by amaschas** [statelyai/xstate#4607] [Q&A]
Hey all, I'm having some trouble figuring out how `invoke` is supposed to be used with anything other than `fromPromise`, since all of the examples only seems to use promises, and the documentation for everything other than `fromPromise` is still pending. What I want to do is fairly straightforward, just pass some context data into a function, perform an operation, then move onto the next state when done, but I can't seem to figure out how to do this, or even which type of actor I should use. Basically what I want to do is this:

```
invoke: {
    src: (data) => {
        // Perform a synchronous operation with the data from input
    },
    input: ({ context }) => ({
        // Get some data from context
    }),
    onDone: {
        target: 'nextState'
    }
}
```

**Comment by davidkpiano** [comment]
Apologies, I will add documentation for the rest of the actor logic creators soon.

You should use `fromPromise` for this use-case too, though. All actors should be considered conceptually async:

```ts
invoke: {
  src: fromPromise(async ({ input }) => {
    const result = processData(input.data);
    return result;
  }),
  input: ({ context }) => ({
    data: context.someData
  }),
  onDone: {
    target: 'nextState',
    actions: assign({
      result: ({ event }) => event.output
    })
  }
});
```

**Comment by amaschas** [comment]
Ah interesting. So there's no way to use a purely synchronous, procedural approach? I'm trying to prototype a state machine that will eventually run on an embedded device, so asynchronous behaviors would literally be impossible in the target machine. Is there an alternative to actors that is purely synchronous?

**Reply by davidkpiano** [reply]
You can do the processing in the `assign` action:

```ts
on: {
  someEvent: {
    actions: assign({
      result: ({ event }) => {
        const result = doSomeProcessing(event.data);
        return result;
// ...
```

**Comment by amaschas** [comment]
Excellent, thank you!

I would just not, I'm probably not the only person attempting to use xstate for this purpose, so an entirely synchronous option for actors might be a nice to have in a future version. Maybe I'll make a feature request?

**Reply by nightire** [reply]
What about using `fromCallback`? The idea is to run a callback actor at the very beginning and perform a synchronous operation when receiving an event that you send to it, and finally send an event back after finishing the sync operations:

```ts
{
  id: 'demo',
  invoke: {
    src: fromCallback(({ receive, sendBack }) => {
      receive((event) => {
        if (event.type === "asking for sync op") {
          const result = doSyncOperation();
          sendBack({ type: "sync op done", data: result })
        }
      });
    })
  },
  on: {
    "sync op done": {
      actions: "doWhateverWithReturnedData",
      target: "next state"
    }
  }
}
```

**Reply by Andarist** [reply]
`fromCallback` used like this might be the easiest way to do this today. You don't even have to use `receive` - just use `sendBack` immediately, it should work.

**Reply by nightire** [reply]
Yes, you're right. I use receive to demonstrate that we can decide "when" to execute the sync ops.

**Reply by nightire** [reply]
Yes, you're right. I use receive to demonstrate that we can decide "when" to execute the sync ops.
